### ISSUE TITLE
changing the acknowledgement interface

### ISSUE THREAD

**Issue by Geal** [streamnative/pulsar-rs#80]
while exploring #51 and #64, and looking at other clients ([Java](http://pulsar.apache.org/api/client/2.5.0-SNAPSHOT/org/apache/pulsar/client/api/Consumer), [Go](https://godoc.org/github.com/apache/pulsar-client-go/pulsar#Consumer), [C++](http://pulsar.apache.org/api/cpp/2.5.0-SNAPSHOT/classpulsar_1_1_consumer.html)), I realized the current acknowledgement system cannot work properly.

If we want to close a consumer, it will have no impact on currently existing `Ack` instances, and if we use [acknowledgement timeout](http://pulsar.apache.org/docs/en/concepts-messaging/#acknowledgement-timeout), we might even try to trigger redelivering on a consumer that does not exist. Also, calling `close_consumer()` from the `Drop` implementation apparently does not work in all cases.

# Proposal

## Move acknowledgement methods to Consumer
Instead of having the `ack()`, `cumulative_ack()` and `nack()` methods on `Ack`, it would make more sense to put those methods on the consumer, as is done in clients from other languages. so we'd write `consumer.acknowledge(&message).await`, instead of the current `Ack::ack()` method which is not async

## Send Ack and Nack directly from Consumer methods when possible

The current `AckHandler` implementation is complex, because it implements [acknowledgement timeout](http://pulsar.apache.org/docs/en/concepts-messaging/#acknowledgement-timeout) (which is discouraged in the documentation). It could be simplified by sending acks directly from `Consumer::ack()`, and nacks directly from `Consumer::nack()` if we're not using the timeout. The `AckHandler` would only be used if we require acknowledgement timeout.

## Properly closing the consumer

Async drop is a [known problem](https://internals.rust-lang.org/t/asynchronous-destructors/11127), but we can have a solution here:
- the consumer holds a `oneshot::Sender<()>`
- we spawn an async block that holds a clone of the `Arc<Connection>` of the consumer, and awaits on the corresponding `Receiver`
- once the consumer is dropped, the sender is dropped, the receiver returns `Err(Canceled)`
- the async block will then call `close_consumer()`

ping @stearnsc any opinions on this?

**Comment by Geal** [comment]
I'm starting multiple PRs on this, to see how the interface would feel

**Comment by Geal** [comment]
I'm running into another issue: it is not possible to send acks asynchronously, because there's no answer to that message: `ConnectionSender` has a channel to `connection::Sender`, and then the sender will receive from that channel and send message on the TCP connection. There's no way to await until the message is actually sent, since we can either call `unbounded_send` on the channel (sync interface) or using `SinkExt::send` but it requires mutable access (it should be possible to make it mutable).
Right now, for most messages, awaiting is done by sending the message with a request key, giving a oneshot sender to the `connection::Receiver` instance, and waiting on that oneshot channel.

Since `connection::Sender` is mainly wrapping the connection's sink, it should be possible to remove it and directly hold the sink in the `ConnectionSender`

**Comment by Geal** [comment]
ok, so, after more investigation, there's an issue with the `UnboundedSender` in `Ack` that is used to communicate with the `Ackhandler`: it does not detect that the `AckHandler` dropped the receiver, so calling `Ack::ack()` does not fail. Closing the channel from the sender side before calling `close_consumer()` is enough to make `ack()` fail as expected

**Comment by Geal** [comment]
with #82 it gets more manageable: acknowledgements are done only through the consumer, so we cannot try to ack a message after the consumer was dropped.

Possible issues:
- it is not possible to acknowledge messages after converting the consumer to a stream (with method like `take()`, etc). I believe the main case will be in `while let Some(message) = consumer.next.await { }`, but still, we should not forbid other patterns. For cases like this, we could add options to either:
  - automatically acknowledge messages as they're taken from the stream
  - send a cumulative ack for all previously seen messages when the consumer drops
  - provide a separate object that could acknowledge messages, like the previous `Ack`, but linked to the `Consumer`, not to the message
- acks are not batched now, but I'm planning to add an ack methods that take a list of messages or message ids to acknlowedge all of them at once

**Comment by Geal** [comment]
done in 1.0 release
